# Credit Card Fraud Detection

# EDA

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


In [2]:
df = pd.read_csv('creditcard.csv')

In [3]:
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...         V21       V22       V23       V24  \
0  0.098698  0.363787  ...   -0.018307  0.277838 -0.110474  0.066928   
1  0.085102 -0.255425  ...   -0.225775 -0.638672  0.101288 -0.339846   
2  0.247676 -1.514654  ...    0.247998  0.771679  0.909412 -0.689281   
3  0.377436 -1.387024  ...   -0.108300  0.005274 -0.190321 -1.175575   
4 -0.270533  0.817739  ...   -0.009431  0.798278 -0.137458  0.141267   

        V25       V26       V27       V28  Amount  Class  
0  0.128539 -0.189115  0.133558 -0.021053  149.62      0  
1  0.167170  0.125895 -0.008983  0.014724    2.69      0  
2 -0.327642 -0.139097 -0.055353 -0.059752  378.66      0  
3  0.647376 -0.221929  0.062723  0.061458  123.50      0  
4 -0.206010  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [4]:
df.shape

(284807, 31)

Observation: <br>
1)There are 284807 data points i.e this many no.of transactions<br>
2)There are 31 Features i.e 31 dimensions

## Description

Time -> Its the seconds elapsed between each transaction and the first transaction in the dataset
<br>
Class: 1 -> fraud, 0 -> no harm

## Analysis

In [5]:
df['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

Observation: there are 492 fraud transactions

In [6]:
percentageOfFraud = 492/(492+284315)
percentageOfFraud*100

0.1727485630620034

Note: 0.17% of transactions are  Fraud

### Creating Sample of 100 with 0.17% transactions of type class 1

In [7]:
s1 = df[df['Class']==0]
sampleClass1=s1[0:500];
sampleClass1.shape

(500, 31)

In [8]:
s2 = df[df['Class']==1]
sampleClass2=s2[0:10];
sampleClass2.shape

(10, 31)

In [9]:
sample = sampleClass1.append(sampleClass2)
sample.shape

(510, 31)

#### Cosine between row vectors

In [10]:
from sklearn.metrics.pairwise import cosine_similarity
d = sample.as_matrix()
cosine_similarity(d[0:1], d[9:10])


array([[0.36180389]])

In [13]:
lst=[]

for i in range(1,510):
    j=i+1
    #print(cosine_similarity(d[0:1], d[i:j])[0][0])
    lst.append(cosine_similarity(d[9:10], d[i:j])[0][0])
lst

[0.36603281633740103,
 0.3639181658838592,
 0.371864955044794,
 0.39129577447494884,
 0.7529411966189501,
 0.7174293174877497,
 0.513884567924309,
 0.4272199257098043,
 1.0,
 0.8512160483640495,
 0.8331928142380813,
 0.4318904748011022,
 0.6536529499115921,
 0.533630019131336,
 0.8149738657557347,
 0.8520343797586534,
 0.9389146770950092,
 0.5836419094742739,
 0.9120768305390461,
 0.4213203529252623,
 0.7089907877735792,
 0.8911100635995834,
 0.8139535570865415,
 0.8825218493732097,
 0.8412818888703114,
 0.7464898822632988,
 0.920389527386811,
 0.8067052788240257,
 0.9426766139677938,
 0.9251620087893493,
 0.9221569956954891,
 0.9507336560513724,
 0.9507336560513724,
 0.9140413722112838,
 0.9140413722112838,
 0.8586037840782367,
 0.9096860166599998,
 0.9176189337513427,
 0.9407378494406858,
 0.8816862074905919,
 0.9090201645655401,
 0.9592958043900317,
 0.9599891143814899,
 0.9594929019613643,
 0.9398010057063506,
 0.950659138580464,
 0.9490272878470489,
 0.8993592597358746,
 0.9496226

In [14]:
li =[]
for i in range(len(lst)):
    li.append(lst[i])
li

[0.36603281633740103,
 0.3639181658838592,
 0.371864955044794,
 0.39129577447494884,
 0.7529411966189501,
 0.7174293174877497,
 0.513884567924309,
 0.4272199257098043,
 1.0,
 0.8512160483640495,
 0.8331928142380813,
 0.4318904748011022,
 0.6536529499115921,
 0.533630019131336,
 0.8149738657557347,
 0.8520343797586534,
 0.9389146770950092,
 0.5836419094742739,
 0.9120768305390461,
 0.4213203529252623,
 0.7089907877735792,
 0.8911100635995834,
 0.8139535570865415,
 0.8825218493732097,
 0.8412818888703114,
 0.7464898822632988,
 0.920389527386811,
 0.8067052788240257,
 0.9426766139677938,
 0.9251620087893493,
 0.9221569956954891,
 0.9507336560513724,
 0.9507336560513724,
 0.9140413722112838,
 0.9140413722112838,
 0.8586037840782367,
 0.9096860166599998,
 0.9176189337513427,
 0.9407378494406858,
 0.8816862074905919,
 0.9090201645655401,
 0.9592958043900317,
 0.9599891143814899,
 0.9594929019613643,
 0.9398010057063506,
 0.950659138580464,
 0.9490272878470489,
 0.8993592597358746,
 0.9496226

In [15]:
#li

In [16]:
dic = dict()


In [17]:
for i in range(0,len(lst)):
    dic[lst[i]]=(i,d[i][30])
dic[0.36603281633740103]

(0, 0.0)

In [18]:
lst.sort()
lst



[0.3639181658838592,
 0.36603281633740103,
 0.371864955044794,
 0.3841837890388966,
 0.3851285635868402,
 0.39129577447494884,
 0.40669165353361103,
 0.4213203529252623,
 0.4272199257098043,
 0.4318904748011022,
 0.44350431430150306,
 0.44736709390344853,
 0.4638974074441167,
 0.48870063991072815,
 0.5079570412124377,
 0.513884567924309,
 0.5226760972866722,
 0.533630019131336,
 0.5503172715506419,
 0.5606539413702267,
 0.5629053639110738,
 0.5817478449607938,
 0.5836419094742739,
 0.5937937272504822,
 0.6092915794432842,
 0.6174514154501195,
 0.6279836194220687,
 0.6432055547360813,
 0.6536529499115921,
 0.6630329824470369,
 0.6648327763689162,
 0.6707354081153399,
 0.6743493938112526,
 0.7051257739122287,
 0.7089907877735792,
 0.7174293174877497,
 0.7397842262672587,
 0.7415858908618538,
 0.7416883930382965,
 0.7464898822632988,
 0.7515569964052831,
 0.7529411966189501,
 0.7544449326054499,
 0.7615112570438488,
 0.7665635045950565,
 0.7716596004495273,
 0.7778723598800379,
 0.7918701

In [19]:
print(dic[0.36603281633740103])
for i in range(10):
    print("(Transaction id, class) -> ",dic[lst[i]]," Similarity: ",lst[i])

(0, 0.0)
(Transaction id, class) ->  (1, 0.0)  Similarity:  0.3639181658838592
(Transaction id, class) ->  (0, 0.0)  Similarity:  0.36603281633740103
(Transaction id, class) ->  (2, 0.0)  Similarity:  0.371864955044794
(Transaction id, class) ->  (50, 0.0)  Similarity:  0.3841837890388966
(Transaction id, class) ->  (163, 0.0)  Similarity:  0.3851285635868402
(Transaction id, class) ->  (3, 0.0)  Similarity:  0.39129577447494884
(Transaction id, class) ->  (88, 0.0)  Similarity:  0.40669165353361103
(Transaction id, class) ->  (19, 0.0)  Similarity:  0.4213203529252623
(Transaction id, class) ->  (7, 0.0)  Similarity:  0.4272199257098043
(Transaction id, class) ->  (11, 0.0)  Similarity:  0.4318904748011022


In [20]:
a = np.empty(5)
a[0]=1
a[1]=2
a

array([1., 2., 0., 0., 0.])